File: epi_judge_cpp/offline_sampling.cc
#include <algorithm>
#include <functional>
#include <iterator>
#include <vector>

#include "test_framework/random_sequence_checker.h"
#include "test_framework/test_timer.h"

using std::bind;
using std::vector;

void RandomSampling(int k, vector<int>* A_ptr) {
  // Implement this placeholder.
  return;
}

bool RandomSamplingRunner(TestTimer& timer, int k, vector<int> A) {
  vector<vector<int>> results;
  timer.Start();
  for (int i = 0; i < 100000; ++i) {
    RandomSampling(k, &A);
    results.emplace_back(begin(A), begin(A) + k);
  }
  timer.Stop();

  int total_possible_outcomes = BinomialCoefficient(A.size(), k);
  sort(begin(A), end(A));
  vector<vector<int>> combinations;
  for (int i = 0; i < BinomialCoefficient(A.size(), k); ++i) {
    combinations.emplace_back(ComputeCombinationIdx(A, A.size(), k, i));
  }
  vector<int> sequence;
  for (vector<int> result : results) {
    sort(begin(result), end(result));
    sequence.emplace_back(
        distance(begin(combinations),
                 find(begin(combinations), end(combinations), result)));
  }
  return CheckSequenceIsUniformlyRandom(sequence, total_possible_outcomes,
                                        0.01);
}

void RandomSamplingWrapper(TestTimer& timer, int k, const vector<int>& A) {
  RunFuncWithRetries(bind(RandomSamplingRunner, timer, k, A));
}

#include "test_framework/test_utils_generic_main.h"

int main(int argc, char* argv[]) {
  generic_test_main(argc, argv, "offline_sampling.tsv", &RandomSamplingWrapper);
  return 0;
}

File: epi_judge_python/offline_sampling.py
import functools

from test_framework.random_sequence_checker import \
    check_sequence_is_uniformly_random, binomial_coefficient, \
    compute_combination_idx, run_func_with_retries
from test_framework.test_utils import enable_timer_hook


def random_sampling(k, A):
    # Implement this placeholder.
    return


@enable_timer_hook
def random_sampling_wrapper(timer, k, A):
    def random_sampling_runner(timer, k, A):
        result = []
        timer.start()
        for _ in range(100000):
            random_sampling(k, A)
            result.append(A[:k])
        timer.stop()

        total_possible_outcomes = binomial_coefficient(len(A), k)
        A = sorted(A)
        comb_to_idx = {
            tuple(compute_combination_idx(A, len(A), k, i)): i
            for i in range(binomial_coefficient(len(A), k))
        }

        return check_sequence_is_uniformly_random(
            [comb_to_idx[tuple(sorted(a))] for a in result],
            total_possible_outcomes, 0.01)

    run_func_with_retries(
        functools.partial(random_sampling_runner, timer, k, A))


from test_framework import generic_test, test_utils

if __name__ == '__main__':
    generic_test.generic_test_main('offline_sampling.tsv',
                                   random_sampling_wrapper)
